src/Compilers/Core/MSBuildTask/Csi.cs

```diff
@@ -0,0 +1,252 @@
+            // ran into, which is described in an email thread attached to bug VSWhidbey 146883.
+            // See also bugs 177762 and 118307 for additional bugs related to response file position.
+            if (ResponseFiles != null)
+            {
+                foreach (var response in ResponseFiles)
+                {
+                    commandLine.AppendSwitchIfNotNull("@", response.ItemSpec);
+                }
+            }
+        }
+        #endregion
+
+        private void AddReferencesToCommandLine(CommandLineBuilderExtension commandLine)
+        {
+            if (References != null)
+            {
+                // Loop through all the references passed in.  We'll be adding separate
+                // /reference: switches for each reference, and in some cases even multiple
+                // /reference: switches per reference.
+                foreach (var reference in References)
+                {
+                    // See if there was an "Alias" attribute on the reference.
+                    string aliasString = reference.GetMetadata("Aliases");
+                    string switchName = "/reference:";
+
+                    if ((aliasString == null) || (aliasString.Length == 0))
+                    {
+                        // If there was no "Alias" attribute, just add this as a global reference.
+                        commandLine.AppendSwitchIfNotNull(switchName, reference.ItemSpec);
+                    }
+                    else
+                    {
+                        // If there was an "Alias" attribute, it contains a comma-separated list
+                        // of aliases to use for this reference.  For each one of those aliases,
+                        // we're going to add a separate /reference: switch to the csc.exe
+                        // command-line
+                        string[] aliases = aliasString.Split(',');
+
+                        foreach (string alias in aliases)
+                        {
+                            // Trim whitespace.
+                            string trimmedAlias = alias.Trim();
+
+                            if (alias.Length == 0)
+                            {
+                                continue;
+                            }
+
+                            // The alias should be a valid C# identifier.  Therefore it cannot
+                            // contain comma, space, semicolon, or double-quote.  Let's check for
+                            // the existence of those characters right here, and bail immediately
+                            // if any are present.  There are a whole bunch of other characters
+                            // that are not allowed in a C# identifier, but we'll just let csi.exe
+                            // error out on those.  The ones we're checking for here are the ones
+                            // that could seriously screw up the command-line parsing or could
```

Comment:
Most recent changes share the code. Someone more knowledgeable on the parsing should probably update the comment